### ISSUE TITLE
Find file search text not saved when cleared

### ISSUE THREAD

**Issue by rohitab** [FarGroup/FarManager#681]
### Far Manager version

3.0.6136.0 x64

### OS version

10.0.22624.1690

### Steps to reproduce

1. Press `Alt+F7` to launch _Find file_
2. Press `Alt+T` to switch to _Text_ mode
3. Type in `hello` in the _Containing_ input box
4. Press `Enter` to start the search
5. Press `Esc` to close _Find file_ results
6. Press `Alt+F7` to re-launch _Find file_
7. Note that the _Containing_ input box shows the previous search `hello`
8. Press `Alt+T`, then `Del` to clear the search text
9. Press `Enter` to start the search
10. Press `Esc` to close _Find file_ results
11. Press `Alt+F7` to re-launch _Find file_

### Expected behavior

The _Containing_ input box should be empty.

### Actual behavior

The _Containing_ input box shows the previous non-empty search text `hello`.

Open the _Editor_, press `F7` to search, press `Del` to clear out the `hello` text from the _Search for_ input box, press `Enter` to search. Now close the _Editor_ and press `Alt+F7` to launch _Find file_ again. You'll notice the _Containing_ input box is now empty.

Appears to be a recent issue which is not present in version `3.0.6134.0 x64`.

**Comment by rohitab** [comment]
Might be an issue with this piece of code that only saves if the string is not empty.

https://github.com/FarGroup/FarManager/blob/c085d23040482e25472d631688a84f0150d40a74/far/findfile.cpp#L3209-L3212

It seems that the Editor saves it regardless of its content.

https://github.com/FarGroup/FarManager/blob/c085d23040482e25472d631688a84f0150d40a74/far/editor.cpp#L220-L223

**Comment by MKadaner** [comment]
Thank you for the report, @rohitab.

I guess I know what happened here. Originally, the similar check was probably intended to avoid modifying the global search options if the user searched for files by name without regard to the content. The thinking went along the lines of, "the user did not supply search pattern, so all search options, like case sensitive, whole words, etc. were not used and should not affect the global settings even if the user accidentally flipped some checkboxes." The behavior was as if the user cancelled the regular Search dialog.

However, before my change, the search pattern was eventually saved to the find-file-specific static buffer _unconditionally_, so the empty search string would be picked up the next time the user opens find file dialog.

I need to think how to imitate old behavior, which is probably expected, without re-introducing yet another piece of global state.

@alabuzhev please assign this one to me and chime in if you have any ideas. Thank you in advance.

**Comment by rohitab** [comment]
Sure thing. I figured as much. Just looking at the replaced code, it's hard to tell that it would introduce this issue. I wouldn't have found it either during a code review.

Just speaking for myself here. I'm fine with clearing the search string, i.e., _not_ imitating the old behavior where the search string in the Editor/Viewer is not cleared. Since everything else is shared between the searches, it makes sense to share the search string as well. Of course, other users might disagree and expect the old behavior.

**Comment by alabuzhev** [comment]
An empty search string does not make any sense in Viewer & Editor.
We probably should just forbid that by disabling Prev/Next/All buttons in the dialog (see Windows Notepad for example).

In Find File, however, it does. Moreover, *conceptually* empty/non-empty is not about the string value per se, but about using or not using the content search, i.e. it is sort of a weird checkbox. If the user cleared the text to not look in files' data, we definitely do not want to restore it on the next invocation.

To summarize:
1. Viewer, Editor and Help search dialogs should not allow persisting empty search strings at all.
2. Find File should, but this "emptiness" should stay within Find File and should not propagate into the global state. Pseudocode:

```C++
bool Config.FindFile.ContentSeach;

...

FindFileDialog.OnCreate():
    FindFileDialog.ContainingText = Config.FindFile.ContentSeach? GlobalSearchString : "";
    
FindFileDialog.OnClose():
    Config.FindFile.ContentSeach = !FindFileDialog.ContainingText.empty();
    if (ContentSeach)
        GlobalSearchString = FindFileDialog.ContainingText;
```

3. All other content-related options (CS, WW, Regex...) in Find File could either be persisted unconditionally (if the user "accidentally flipped some checkboxes", they can always flip them back) or only when the string was not empty. I don't have a strong preference here.

**Comment by MKadaner** [comment]
Thank you for the exact spec, @alabuzhev. I like the behavior you suggested; will fix it as soon as I can. Quick question:

> bool Config.FindFile.ContentSeach;

Do I read it correctly: you want this flag to be a config setting persisting across Far sessions?

**Comment by MKadaner** [comment]
> We probably should just forbid that by disabling Prev/Next/All buttons in the dialog (see Windows Notepad for example).

It does not seem to be a common behavior in Far though. In other dialogs (all of them?), where an empty string does not make sense (for example _Make folder_), the `OK` button is not disabled if the string becomes empty. (In fact, I tried to find a place to copy from and could not). Are you sure we want this behavior?

> Viewer, Editor and Help search dialogs should not allow persisting empty search strings at all.

That can be done easily. What about persisting the options in this case? I would still save them. Then, the clients can do whatever they see fit. Viewer and Editor just bail out of the search if the patter is empty, while Help lists all topics, which makes sense.

**Comment by pasha-zzz** [comment]
> I guess I know what happened here

Sync Viewer/Editor search with Find File search is very bad idea at all. More often Find File by name only needed. But search phrase from E/V fills search text string when I forgot remove it and no files may be found. Find File search text must be empty by default ALWAYS.

**Comment by yegor-mialyk** [comment]
I would agree that automatically populating the 'Containing Text' field in the 'Find File' dialog from the editor is probably not a good idea.

**Comment by pasha-zzz** [comment]
> The flow from 'Find File' to Editor/Viewer is pretty much acceptable, but the flow from Editor/Viewer to 'Find File' is questionable.

Absolutely agree

**Comment by MKadaner** [comment]
> The flow from 'Find File' to Editor/Viewer is pretty much acceptable

Not just "acceptable" but very desirable, IMO. I use it every day.

As for picking Find File search string from Viewer / Editor... Well, I introduced this behavior in 6134. It seemed to be agreed upon. I slightly prefer it to starting from the empty search pattern, but I have lived with the old behavior for a long time; one can always pull the search pattern from the history.

More opinions? @alabuzhev? @HamRusTal?

**Comment by alabuzhev** [comment]
> It does not seem to be a common behavior in Far though.

I don't think it's a deliberate design decision, but rather for historical reasons.
Initially dialogs were quite simple and without any dynamic modifications in DialogProc, because there was no DialogProc in the first place.

> Are you sure we want this behavior?

Can't see why not. However, if you prefer not to do it now, it's fine. Checking that the string is empty after closing the dialog and exiting early should do for now. We can review all the dialogs and add prechecks later.

> It seemed to be agreed upon.

Or overlooked.
It is natural to propagate the string and other options from Find File to Editor/Viewer: if the user searches in files, they will likely view/edit them from the search results or tmppanel and look for the same string.

Editor/Viewer -> Find File is questionable indeed. I can imagine scenarios like "I opened a file in Editor, looked for a string, didn't find it, maybe it's in another file, Alt+F7", but I can also imagine scenarios where a subsequent Find File is completely unrelated to previous views/edits.

**Comment by HamRusTal** [comment]
IMHO __passing the search text from Find File to Editor/Viewer Search is vital__, for the scenario described [above](https://github.com/FarGroup/FarManager/issues/681#issuecomment-1545374133):

> if the user searches in files, they will likely view/edit them from the search results or tmppanel and look for the same string

If the text is empty, the E/V Search should pick the last non-empty text/regex though, for two reasons: (1) the empty text makes no sense; (2) it is useful sometimes to be able to search e.g. by mask only and then repeat the pre-FF E/V search in the found files.

As for the opposite direction (E/V Search to FF), I'm inclined to agree with those who oppose this. It's much more often that I need an empty Containing Text field in FF than pre-filled from the preceding E/V Search. On the rare occasions when I need the E/V Search text in FF, it's no problem to pick it from the drop-down history.

So, I agree with [this](https://github.com/FarGroup/FarManager/issues/681#issuecomment-1545141577):

> Find File search text must be empty by default ALWAYS.

**Comment by MKadaner** [comment]
Thanks for the input, I'll fix the behavior accordingly.

**Comment by HamRusTal** [comment]
An interesting point to discuss is whether Find File and E/V Search should share the same history. On the one hand, it may be useful for the scenario outlined in my earlier [message](https://github.com/FarGroup/FarManager/issues/681#issuecomment-1545743340):

> to be able to search e.g. by mask only and then repeat the pre-FF E/V search in the found files

On the other hand, it may fill the Find File Text history with items relevant only to searching within particular files (being edited/viewed) but not to searching across directories.

Ideally, we should elaborate some smart tradeoff. I'll think of it myself and I encourage others the do the same.

What I can suggest right now is that the histories are distinct, but (some?) items are copied between them. This will let users at least remove different items from these histories.

**Comment by MKadaner** [comment]
Personally, I am absolutely for making E/V history available in FF. I use it very often, maybe every day. A typical scenario goes like this: I open a file I think contains what I am looking for, search, do not find it, and switch to FF to search similar files around for the same pattern (with the same search options). If the search pattern from E/V is not copied to FF as requested above by many people, at least the history should be there for me.

Let's don't change this behavior for now and dicuss it separately.

**Comment by rohitab** [comment]
> So, I agree with [this](https://github.com/FarGroup/FarManager/issues/681#issuecomment-1545141577):
> 
> > Find File search text must be empty by default ALWAYS.

Why shouldn't it just use text from the previous file search, like it always has? Say I'm searching for a file containing some text or hex bytes and I don't find it. I go to a different folder to repeat the search. With the change you're suggesting, I would need to re-enter or select the search string or bytes from history for each search. This can be a real hassle if you have to search through multiple folders and initialize search each time.

@MKadaner I know you're probably still working on the changes, but I went ahead and tried out [`MKadaner:mzk-find-file-search-string`](https://github.com/MKadaner/FarManager/tree/mzk-find-file-search-string), just to see how it would work. I thought it would be good to provide some feedback before the changes are finalized.

### Example

Say, I want to search for the bytes `72 6F 68 69 74 61 62` in executable files in multiple folders. Here is what I would do:

1. Press `Alt+F7` to open _Find file_
2. Type in `<exec>` for the _file mask_
3. Press `Alt+H` to switch to _Hex_ mode
4. Type in `72 6F 68 69 74 61 62` in the _Containing_ input box
5. Press `Enter` to search

Now, let's say I want to repeat the above search in different folder(s). Here is what I need to do with the new (suggested) changes:

1. Press `Alt+F7` to do reopen _Find file_
2. The _file mask_ is already filled in with `exec`, so that's good
3. The _Containing_ input box is empty and the mode has switched to _Text_
4. Press `Alt+H` to switch to _Hex_ mode
5. There is no history available, so type the search bytes `72 6F 68 69 74 61 62` again
6. Even if you press `Alt+T` to switch to _Text_ mode, the previously searched bytes are not available in the history, even as a text string
7. Press `Enter` to search
8. Repeat Steps 1 through 7 for each new folder

With the current version of _Far_, all I need to do is press `Alt+F7` followed by `Enter` to repeat the search in each new folder, which makes it a lot easier, and in my opinion, the way it should continue to work.

### Suggestion

How about having a separate checkbox for _Content Search_, instead of basing it on whether the search text is empty or not? That way, if someone is searching for just files, the search text will not interfere with the search, regardless of whether it came from Editor/Viewer or from a previous search. And if someone is searching for content, the previous search text (or bytes) will be available to be reused.

**Comment by HamRusTal** [comment]
> Why shouldn't it just use text from the previous file search, like it always has?

Fair point. Somehow I missed that the traditional behavior is to reuse the last searched text. Probably because I use Find File not often enough.

> How about having a separate checkbox for _Content Search_, instead of basing it on whether the search text is empty or not?

It's an idea I had too, also hinted by an earlier [message](https://github.com/FarGroup/FarManager/issues/681#issuecomment-1540482187):

> _conceptually_ empty/non-empty is not about the string value per se, but about using or not using the content search, i.e. it is sort of a weird checkbox

Yet I consider it a bit out of scope here. It's worth a separate issue and IMHO should be implemented after the current behavior is stabilized.

**Comment by MKadaner** [comment]
Thank you for the early feedback, @rohitab, always appreciated.

OK. Let's get back to basics. The original complaint here was that the empty search pattern is not saved across FF invocations. I am going to implement the following:
- Add Find-File-specific persistent search options which include flags (CS, WW, Fuzzy).
- On close, copy all options, both search pattern and flags, to the E/V persistent instance if the search pattern is not empty.

![image](https://github.com/FarGroup/FarManager/assets/35671029/8c3cb147-96e5-430c-858e-dcb374413c1c)
Sold!

**Comment by rohitab** [comment]
> It's an idea I had too, also hinted by an earlier [message](https://github.com/FarGroup/FarManager/issues/681#issuecomment-1540482187):

Yes, in fact _Content Search_ was briefly discussed in #601 as well. See https://github.com/FarGroup/FarManager/issues/601#issuecomment-1374921006 and https://github.com/FarGroup/FarManager/issues/601#issuecomment-1374925576.

> Yet I consider it a bit out of scope here. It's worth a separate issue and IMHO should be implemented after the current behavior is stabilized.

I agree. It's probably best to just resolve the issue in this ticket for now, and discuss any other changes separately, giving people time to provide feedback.

**Comment by MKadaner** [comment]
Fixed in #686.

**Comment by rohitab** [comment]
I just tested version `3.0.6144.0 x64`. The issue is resolved. Thank you.

I'm closing this ticket since the original issue has been fixed. If required, please create separate tickets for any other items that were discussed, but not yet implemented, or use [Discussions](https://github.com/FarGroup/FarManager/discussions).
